### DISCUSSION TITLE
Looping an API query

### DISCUSSION THREAD

**Discussion by JClarke90** [Budibase/budibase#11803] [Help] [answered]
Thanks for answering my other questions, I have one more that is more complex than the others.

I want to be able to run an API query for each row in a table and then concat all of the results and display that in another table, no idea if this is possible, maybe through automation but hopefully someone will know.

Example:

For each row in this table, take the ID from it, and run api queries for each one, and then combine the results together in another table, assume there will be more than 2 results.

![image](https://github.com/Budibase/budibase/assets/12259768/c866550b-6f08-4684-8e69-7def772e8aa7)

I can probably figure out how to combine the results but I don't know how to look the query.

**Comment by ConorWebb96** [comment]
Hey @JClarke90,

I think what you are asking is doable there are quite a few ways you could go about this. E.g. you could setup a [cron job](https://docs.budibase.com/docs/cronjob) automation that would run daily (Daily is only an example can be changed to whatever you want). You may need to amend your already existing tables, as you wouldn't want this to constantly [query](https://docs.budibase.com/docs/query-rows)/run over records were the data has been fetched and saved. (Include a boolean column called synced the name doesn't matter, set to false initially once records have been synced/saved update the record to true.) Doing this would allow you to get all records that haven't synced and then save their related subscriptions etc using various steps such as; external data connector step (for your rest queries), update/create steps (Modify/create rows) etc.

Additionally, you could use [app action](https://docs.budibase.com/docs/app-action) automation. This can be triggered with the usage of the [trigger automation](https://docs.budibase.com/docs/application-actions#trigger-automations) action, you can attach this to various components within Budibase. For example `on row click`, you'd pass the _id (row id) as well as the subscription id through bindings that you'd create when setting up the [app action](https://docs.budibase.com/docs/app-action) automation. The _id would be used to create a relationship between the clicked record and the subscription id would be used for the rest query. The results from the rest query would then be used to create the related subscription row.

**Optionally you could do the below**

Instead of saving these records to a table, I think there may be an easier method. Since you mentioned and API query to get the record, you could do a row on click which directs them to a new page `user/subscription/:id` (:id would be the subscription id in this case as i think this is what you'd use for the api call.)  It doesn't have to be a new page, you could run this and display the related contents in a side panel instead.

On this new page you can click the `on screen load` actions. Attach 2 new actions to this; [execute query ](https://docs.budibase.com/docs/data-actions#execute-query) configure it to use your `get record` rest query, use the id you passed through the url. Below that use the [update state](https://docs.budibase.com/docs/data-actions#update-state) action and configure this to use the `{{ Action 1.Query result }}` - results of your query.

<img width="889" alt="Screenshot 2023-09-19 at 12 30 40" src="https://github.com/Budibase/budibase/assets/126772285/8e57f928-7808-4153-b57b-4e6d5723e0b3">

<img width="798" alt="Screenshot 2023-09-19 at 12 30 54" src="https://github.com/Budibase/budibase/assets/126772285/117b3fa9-bb8e-46c9-b6ae-a85a602e12eb">

From here you could then look at binding up and appending the page however you like. The only downside to this is the rest query would get run each time a user clicks on a record.

<img width="1105" alt="Screenshot 2023-09-19 at 12 31 11" src="https://github.com/Budibase/budibase/assets/126772285/91dc7b1b-5755-46ef-8e38-c43850a767de">

I hope this points you in the right direction!

**Comment by JClarke90** [comment]
Ah I probably should have gone into a bit more detail.

The end goal here is to list all VMs from all subscriptions.

So I would run one query to get all the subscriptions, this can be on page load, and the results/table can be hidden, this will only run once when the app loads.

From these results I then need to grab the ID of each row that was returned and run the get VM query using the results.

I’ve managed to get a list of the Subscription ids using automation and {{#forin steps.1.response}} so I could tie this to a button in the app for reload.

But I’m now not sure what to do with the list of ID’s to get the other query to go through them all and then join the response arrays together.

I haven’t tested anything in your other comment as it’s late here but I wanted to get a response just in case it helped to explain in more detail. 

I’ll look into this more in the morning.

**Comment by JClarke90** [comment]
@ConorWebb96 thanks for your help on this it gave me a few ideas and I made some progress.

I now have an automation task that queries subscriptions and uses a forin loop that will pull all of the IDs and then feed them into the second query.

I managed to get them dumped into a Budibase table, as I don’t know if it’s possible to have results from a looping query simply appear in a table in the design view?

If I can’t find a way to get them straight from the looping api into the design view I could have this run every 30 mins and just keep updating the BB table but when I did this, it duplicates the entries and I don’t know how to prevent that, it would need to do a query then a Conditional if match on the subscription Id update the row instead of create a new one. Is this possible?

**Reply by ConorWebb96** [reply, marked as answer]
Hey @JClarke90,

You can use the results from the looped query step as the basis for creating rows. You'd setup looping for your `create row` step and make it loop through the results of your External Data connect.

In this case the loop input type would be set to `array` and the binding/value you'd use would look something like: `{{ steps.1.items }}`. The step number would be what you'd change (Occasionally you'll have to switch items to rows or value depending on the automation step you are using), and from there you should be able to use `{{ loop.currentItem.fieldName }}` to bind them to their relevant fields. I've provided a screenshot example of this below;
<img width="487" alt="Screenshot 2023-09-20 at 14 50 22" src="https://github.com/Budibase/budibase/assets/126772285/cbfc6d15-1377-465e-bb2a-c052257ac0df">

If you go into your automation area and click the run test button you can see the outputs of each step. **(This does run the automation, so if its create rows you may need to delete duplicates as you are testing the automation.)** I've attached a screenshot of this below to hopefully help explain this a little bit better.
<img width="709" alt="Screenshot 2023-09-20 at 14 53 48" src="https://github.com/Budibase/budibase/assets/126772285/aa0fdd88-7f43-4e44-a967-aa292f5be0d4">

What you can do to prevent duplicates/what I have done in the past when handling large data sets. Is to use the `JS SCRIPTING` step,
then write code to compare the rows from your rest api (External Data Connector) to the rows in the `query rows` step. If any match then remove them and output a new array with nothing but unique results that haven't been added yet. From here what you would do is use the results from that `JS SCRIPTING` for looping over and creating the rows. 

If you were just working with a single row, you'd likely be able to just use the `CONDITION` step. But in your case you have a lot of data, so the `JS SCRIPTING` step is your best option in my opinion.

I hope this helps.

**Comment by JClarke90** [comment]
Hi @ConorWebb96, thanks for the info.

I've managed to build a JS Scripting query to only output items that don't already exist in the table, and it outputs an array like below.

![image](https://github.com/Budibase/budibase/assets/120619360/06abafa9-e77f-44e8-a35f-5efde034d474)

```
{
  "success": true,
  "value": [
    {
      "name": "000000",
      "id": "/subscriptions/0000000000000/resourceGroups/0000000/providers/Microsoft.Compute/virtualMachines/000000-Dev"
    },
    {
      "name": "000000000",
      "id": "/subscriptions/00000000000/resourceGroups/0000000000000/providers/Microsoft.Compute/virtualMachines/000000-master"
    }
  ]
}
```

But when I try and use this output to create rows the value is always empty. Not sure if I'm doing something wrong or its a bug but i've tried many different things, with handlebars and JS but its always blank. I know I need to use looping but im just trying to get anything to display at the mo but not having any luck.

![image](https://github.com/Budibase/budibase/assets/120619360/b2b9ea35-2818-4472-98e3-369a29ef0c84)

![image](https://github.com/Budibase/budibase/assets/120619360/e57f6578-59c3-4423-8b3e-c56b91ef959f)

I even just tried dumping the whole array in there but still nothing

![image](https://github.com/Budibase/budibase/assets/120619360/4b278128-d857-4445-b0e2-4198aab424cd)

![image](https://github.com/Budibase/budibase/assets/120619360/41aacc4f-ae89-4ff1-bdbf-b03488560e2a)

**Comment by JClarke90** [comment]
Figured this one out on my own, its a bug with the step numbers for the outputs from JS Scripting.

I have raised it here: https://github.com/Budibase/budibase/issues/11866

**Reply by ConorWebb96** [reply]
Thanks at @JClarke90, yes you are right the automation steps get out of sync whenever you add looping to any step within the automation. I believe there are a few suggested changes for this. E.g. naming your automation steps. 

However, overall it does need fixed. Thanks for creating this issue.
